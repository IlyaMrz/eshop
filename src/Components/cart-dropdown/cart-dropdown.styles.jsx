import styled from 'styled-components';
import CustomButton from '../custom-button/custom-button';

export const CartDropdownButton = styled(CustomButton)`
  margin-top: auto;
`;

export const CartDowndropContainer = styled.div`
    position: absolute;
    width: 240px;
    min-height: 340px;
    max-height: 48vh;
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 20px;
    border: 1px solid black;
    background-color: white;
    top: 90px;
    right: 40px;
    z-index: 5;
    @media screen and (max-width: 400px ) {
        max-height: 70vh;
    }
`;

CartDropdownButton.displayName = 'CartDropdownButton';

export const CartItemsContainer = styled.div`
    width: 90%
    display: flex;
    flex-direction: column;
    overflow: scroll;
    overflow-x: hidden; 
    margin-bottom: 10px;

`;


export const EmptyMessageContainer = styled.span`
font-size: 18px;
margin: 50px auto;
`;

EmptyMessageContainer.displayName = 'EmptyMessageContainer';