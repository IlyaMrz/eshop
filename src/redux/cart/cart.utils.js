export  const addItemToCart = (cartItems, cartItemToAdd) => {
    const existignCartItem = cartItems.find(
        cartItem => cartItem.id === cartItemToAdd.id)
    if (existignCartItem) {
        return cartItems.map(cartItem => 
            cartItem.id === cartItemToAdd.id 
            ? {...cartItem,
                quantity: cartItem.quantity + 1 } 
            : cartItem
            )
    }
    return [...cartItems, {...cartItemToAdd, quantity: 1} ]
}