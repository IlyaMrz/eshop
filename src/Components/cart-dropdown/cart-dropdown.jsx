import React from 'react';
import CartItem from '../cart-item/cart-item';
import CustomButton from '../custom-button/custom-button';
import './cart-dropdown.scss';
import { connect } from 'react-redux';
import { selectCartItems } from '../../redux/cart/cart.selectors';

const CartDowndrop = ({cartItems}) => (
    <div className='cart-dropdown'>
        <div className='cart-items'>
            {
                cartItems.map(cartItem => <CartItem key={cartItem.id} item={cartItem}/>)
            }
            <CustomButton>GO TO CHECKOUT</CustomButton>
        </div>
    </div>
)
const mapStateToProps = (state) => ({
    cartItems: selectCartItems(state)
})

export default connect(mapStateToProps, null)(CartDowndrop);