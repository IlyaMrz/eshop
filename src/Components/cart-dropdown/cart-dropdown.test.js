import React from 'react';
import { shallow } from 'enzyme';

import { CartDowndrop } from './cart-dropdown';
import CartItem from '../cart-item/cart-item';

import { toggleCartHidden } from '../../redux/cart/cart.actions';

describe('CartDowndrop component', () => {
  let wrapper;
  let mockHistory;
  let mockDispatch;
  const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

  beforeEach(() => {
    mockHistory = {
      push: jest.fn()
    };

    mockDispatch = jest.fn();

    const mockProps = {
      cartItems: mockCartItems,
      history: mockHistory,
      dispatch: mockDispatch
    };

    wrapper = shallow(<CartDowndrop {...mockProps} />);
  });

  it('should render CartDowndrop component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call history.push when button is clicked', () => {
    wrapper.find('CartDropdownButton').simulate('click');
    expect(mockHistory.push).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
  });

  it('should render an equal number of CartItem components as the cartItems prop', () => {
    expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
  });

  it('should render EmptyMessageContainer if cartItems is empty', () => {
    const mockProps = {
      cartItems: [],
      history: mockHistory,
      dispatch: mockDispatch
    };

    const newWrapper = shallow(<CartDowndrop {...mockProps} />);
    expect(newWrapper.exists('EmptyMessageContainer')).toBe(true);
    expect(newWrapper).toMatchSnapshot();
  });
});