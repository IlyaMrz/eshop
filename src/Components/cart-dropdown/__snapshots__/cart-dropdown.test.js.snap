// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDowndrop component should render CartDowndrop component 1`] = `
<styled.div>
  <styled.div>
    <CartItem
      item={
        Object {
          "id": 1,
        }
      }
      key="1"
    />
    <CartItem
      item={
        Object {
          "id": 2,
        }
      }
      key="2"
    />
    <CartItem
      item={
        Object {
          "id": 3,
        }
      }
      key="3"
    />
  </styled.div>
  <CartDropdownButton
    onClick={[Function]}
  >
    GO TO CHECKOUT
  </CartDropdownButton>
</styled.div>
`;

exports[`CartDowndrop component should render EmptyMessageContainer if cartItems is empty 1`] = `
<styled.div>
  <styled.div>
    <EmptyMessageContainer>
      Your cart is empty
    </EmptyMessageContainer>
  </styled.div>
  <CartDropdownButton
    onClick={[Function]}
  >
    GO TO CHECKOUT
  </CartDropdownButton>
</styled.div>
`;
